src/eduid_webapp/email/views.py

```diff
@@ -0,0 +1,295 @@
+            'error': {'form': 'out_of_sync'}
+        }
+    current_app.logger.info('Email address {!r} made primary '
+                            'for user {!r}'.format(email, user))
+    emails = {'emails': user.mail_addresses.to_list_of_dicts()}
+    return EmailListPayload().dump(emails).data
+
+
+@email_views.route('/verify', methods=['POST'])
+@UnmarshalWith(VerificationCodeSchema)
+@MarshalWith(EmailResponseSchema)
+@require_dashboard_user
+def verify(user, code, email, csrf_token):
+    """
+    """
+    if session.get_csrf_token() != csrf_token:
+        abort(400)
+    current_app.logger.debug('Trying to save email address {!r} as verified '
+                             'for user {!r}'.format(email, user))
+    db = current_app.verifications_db
+    state = db.get_state_by_eppn_and_code(user.eppn, code)
+
+    timeout = current_app.config.get('EMAIL_VERIFICATION_TIMEOUT', 24)
+    if state.is_expired(timeout):
+        msg = "Verification code is expired: {!r}".format(state.verification)
+        current_app.logger.debug(msg)
+        return {
+            '_status': 'error',
+            'error': {'form': 'emails.code_expired'}
+        }
+
+    if email != state.verification.email:
+        msg = "Invalid verification code: {!r}".format(state.verification)
+        current_app.logger.debug(msg)
+        return {
+            '_status': 'error',
+            'error': {'form': 'emails.code_invalid'}
+        }
+
+    state.verification.is_verified = True
+    state.verification.verified_ts = datetime.datetime.now()
+
+    state.verification.verified_by = user.eppn
+
+    current_app.verifications_db.save(state)
+    msg = 'Saved verification code: {!r} '.format(state.verification)
+    current_app.logger.debug(msg)
+
+    other = current_app.dashboard_userdb.get_user_by_mail(email,
+                                                     include_unconfirmed=True)
+    if other and other.mail_addresses.primary and \
+            other.mail_addresses.primary.email == email:
+        # Promote some other verified e-mail address to primary
+        for address in other.mail_addresses.to_list():
+            if address.is_verified and address.email != email:
+                other.mail_addresses.primary = address.email
+                break
+        other.mail_addresses.remove(email)
+        save_dashboard_user(other, dbattr_name='dashboard_userdb')
+        msg = 'Stole email {!r} from user {!r} for user {!r}'.format(email,
+                                                                     user,
+                                                                     other)
+        current_app.logger.info(msg)
+
+    new_email = MailAddress(email = email, application = 'dashboard',
+                            verified = True, primary = False)
+
+    if user.mail_addresses.primary is None:
+        new_email.is_primary = True
+    try:
+        user.mail_addresses.add(new_email)
+    except DuplicateElementViolation:
```

Comment:
When can this happen? Shouldn't the mail address already be unique after the "stealing" code above?